packages/use-optionally-controlled-state/src/index.tsx

```diff
@@ -1,23 +1,31 @@
-import {useState} from 'react';
+import {useState, useCallback, useRef} from 'react';
 import useConstant from 'use-constant';
 
+export type OptionalState<Value> =
+  | {
+      controlledValue?: Value;
+      initialValue: Value;
+      onChange?(value: Value): void;
+    }
+  | {
+      controlledValue: Value;
+      initialValue?: Value;
+      onChange?(value: Value): void;
+    };
+
 /**
  * Enables a component state to be either controlled or uncontrolled.
  */
 export default function useOptionallyControlledState<Value>({
   controlledValue,
   initialValue,
   onChange
-}: {
-  controlledValue?: Value;
-  initialValue?: Value;
-  onChange?(value: Value): void;
-}): [Value | undefined, (value: Value) => void] {
+}: OptionalState<Value>): [Value, (value: Value) => void] {
   const isControlled = controlledValue !== undefined;
   const initialIsControlled = useConstant(() => isControlled);
   const [stateValue, setStateValue] = useState(initialValue);
```

Comment:
Maybe we could add the `Value` generic here to avoid the cast in L48?

Reply:
no, initialValue is optional, I'd rather only cast it when needed. 
The whole idea was to make sure the value doesn't get `| undefined` in the consumer, like:
```
const [value, setValue] = useOptionalState({ initialValue: 1 });
// value was `number | undefined`
```

The new `Options` type enforces that the user specifies either `initialValue` or `controlledValue`.

On the plus side, I replaced the casting with the `!` operator, verified these work:
```tsx
function typeTest(maybe?: number) {
  const [value1, setValue1] = useOptionallyControlledState({ controlledValue: 1 }); // type is number
  const [value2, setValue2] = useOptionallyControlledState({ initialValue: 1 });  // type is number
  const [value3, setValue3] = useOptionallyControlledState({ initialValue: undefined });  // type is undefined
  const [value4, setValue4] = useOptionallyControlledState({ initialValue: undefined });  // type is undefined
  const [value5, setValue5] = useOptionallyControlledState({ initialValue: maybe });  // type is number | undefined
}
```

actually, we can enforce the types not to allow both initialValue and controlledValue, but I think it will cause bugs than value